modules/websearch/lib/search_engine_obelix.py

```diff
@@ -0,0 +1,561 @@
+# -*- coding: utf-8 -*-
+
+## This file is part of Invenio.
+## Copyright (C) 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014 CERN.
+##
+## Invenio is free software; you can redistribute it and/or
+## modify it under the terms of the GNU General Public License as
+## published by the Free Software Foundation; either version 2 of the
+## License, or (at your option) any later version.
+##
+## Invenio is distributed in the hope that it will be useful, but
+## WITHOUT ANY WARRANTY; without even the implied warranty of
+## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+## General Public License for more details.
+##
+## You should have received a copy of the GNU General Public License
+## along with Invenio; if not, write to the Free Software Foundation, Inc.,
+## 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.
+
+# pylint: disable=C0301,W0703
+
+"""Invenio Obelix Search Engine """
+
+import json
+import time
+
+from invenio.errorlib import register_exception
+from invenio.config import CFG_WEBSEARCH_OBELIX_REDIS, CFG_WEBSEARCH_OBELIX_USER_KEY
+
+
+def obelix_get_redis():
+    """
+    Build's redis object based on CFG_WEBSEARCH_OBELIX_REDIS
+    :return:
+    """
+
+    try:
+        from redis import Redis
+
+        return Redis(CFG_WEBSEARCH_OBELIX_REDIS)
+    except Exception:
+        register_exception(alert_admin=True)
+
+
+class ObelixSearchEngine(object):
+    def __init__(self, uid, hitset, redis=None):
+        """
+        ObelixSearchEngine class, built as a class to reuse
+        settings and stored data (recommendations)
+
+        :param uid: the current user logged in
+        :param hitset: the list of records (per collection) that we want to re-rank
+        :param redis: mainly used for mocking, set redis to RedisMock when running tests
+        :return:
+            ObelixSearchEngine object, usage example:
+
+                search_engine = ObelixSearchEngine(uid, records)
+                search_engine.rank() --> returns list of records and scores, sorted recommendations
+                search_engine.rank() --> [1,12,3,4], [0.9,0.8,0.7,0.6]
+
+        """
+
+        self.settings = ObelixSearchEngineSettings(redis)
+        self.uid = uid
+        self.__hitset = hitset
+
+        if redis:
+            self.redis = redis
+        else:
+            self.redis = obelix_get_redis()
```

Comment:
Wouldn't it be better this check is done before instantiating `ObelixSearchEngineSettings so you have `redis` already set and you don't have to check it twice?

Something like
```Python
self.redis = redis or obelix_get_redis()
self.settings = ObelixSearchEngineSettings(redis)
....
```